pyro/infer/renyi_elbo.py

```diff
@@ -0,0 +1,244 @@
+
+                    elbo_particle = elbo_particle - log_prob_sum
+
+            elbo_trace.append(elbo_particle)
+
+        if self.num_particles == 1:
+            elbo = elbo_trace[0]
+        else:
+            if is_vectorized:
+                elbo_particles = elbo_trace[0]
+            else:
+                elbo_particles = torch.tensor(elbo_trace)  # no need to use .new*() here
+
+            elbo_particles_scaled = (1. - self.alpha) * elbo_particles
+            elbo_scaled = log_sum_exp(elbo_particles_scaled, dim=0) - math.log(self.num_particles)
+            elbo = elbo_scaled.sum().item() / (1. - self.alpha)
+
+        loss = -elbo
+        warn_if_nan(loss, "loss")
+        return loss
+
+    def loss_and_grads(self, model, guide, *args, **kwargs):
+        """
+        :returns: returns an estimate of the ELBO
+        :rtype: float
+
+        Computes the ELBO as well as the surrogate ELBO that is used to form the gradient estimator.
+        Performs backward on the latter. Num_particle many samples are used to form the estimators.
+        """
+        elbo_trace = []
+        surrogate_elbo_trace = []
+        tensor_holder = None
+        is_vectorized = self.vectorize_particles and self.num_particles > 1
+
+        # grab a vectorized trace from the generator
+        for model_trace, guide_trace in self._get_traces(model, guide, *args, **kwargs):
+            elbo_particle = 0
+            surrogate_elbo_particle = 0
+
+            # compute elbo and surrogate elbo
+            for name, site in model_trace.nodes.items():
+                if site["type"] == "sample":
+                    if is_vectorized:
+                        log_prob_sum = site["log_prob"].reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+                    elbo_particle = elbo_particle + log_prob_sum.detach()
+                    surrogate_elbo_particle = surrogate_elbo_particle + log_prob_sum
+
+            for name, site in guide_trace.nodes.items():
+                if site["type"] == "sample":
+                    log_prob, score_function_term, entropy_term = site["score_parts"]
+                    if is_vectorized:
+                        log_prob_sum = log_prob.reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+
+                    elbo_particle = elbo_particle - log_prob_sum.detach()
+
+                    if not is_identically_zero(entropy_term):
+                        surrogate_elbo_particle = surrogate_elbo_particle - log_prob_sum
+
+                    if not is_identically_zero(score_function_term):
+                        surrogate_elbo_particle = (surrogate_elbo_particle +
+                                                   (self.alpha / (1. - self.alpha)) * log_prob_sum)
+
+            if is_identically_zero(elbo_particle):
+                if tensor_holder is not None:
```

Comment:
I wonder if we could use `util.Dice` to simplify this instead as in `TraceEnum_ELBO`?

Reply:
hmm..interesting. I'll have to read more on the dice operator to comment.

Reply:
@neerajprad I use `tensor_holder` because I worry that for some particle, elbo_particle might be 0 (for any reason), and for other particles, elbo_particle are tensors. This is pretty cheap I guess.

I did have two containers: elbo_trace and surrogate_trace which stores what you suggested.

Reply:
From what I can see, it should be safe to just consider the vectorized and non vectorized cases separately. In the vectorized case, we'll just have one element in the `elbo_particles` list and in the non-vectorized case each element should be a scalar (so it shouldn't matter if something is a 0).

Reply:
So you mean that we initiate `elbo_particle` with `torch.tensor(0.)` instead of `0.`? How could we know the data type and device then?

Reply:
I see the issue. For the non-vectorized case, you are right that there is a bit of an issue, since we may (rarely, but perhaps due to conditional execution), have a case where we have `elbo_particles` like `[0, torch.tensor(-1.3), torch.tensor(-2.2)]`.